File: src/dere_cli/subcommands.py
"""Typer subcommands for dere CLI."""
from __future__ import annotations

import os
import signal
import subprocess
import sys
import time

import typer

from dere_shared.constants import DEFAULT_DAEMON_URL
from dere_shared.paths import get_config_dir, get_data_dir

typer_app = typer.Typer(help="Dere - Personality-layered wrapper for Claude Code")
daemon_app = typer.Typer(help="Daemon management")
config_app = typer.Typer(help="Configuration management")

typer_app.add_typer(daemon_app, name="daemon")
typer_app.add_typer(config_app, name="config")


@typer_app.callback(invoke_without_command=True)
def main_callback(ctx: typer.Context):
    """Dere - Personality-layered wrapper for Claude Code"""
    if ctx.invoked_subcommand is None:
        print(ctx.get_help())


# Daemon commands
@daemon_app.callback(invoke_without_command=True)
def daemon_callback(ctx: typer.Context):
    """Daemon management"""
    if ctx.invoked_subcommand is None:
        print(ctx.get_help())


@daemon_app.command()
def status():
    """Check daemon status"""
    import httpx

    try:
        with httpx.Client() as client:
            response = client.get(f"{DEFAULT_DAEMON_URL}/health", timeout=2.0)
            if response.status_code == 200:
                data = response.json()
                print("Daemon is running")
                print(f"  Database: {data.get('database', 'unknown')}")
                print(f"  DereGraph: {data.get('dere_graph', 'unknown')}")
            else:
                print("Daemon is not responding correctly")
                sys.exit(1)
    except (httpx.ConnectError, httpx.TimeoutException):
        print("Daemon is not running")
        sys.exit(1)
    except Exception as e:
        print(f"Error checking daemon: {e}", file=sys.stderr)
        sys.exit(1)


@daemon_app.command()
def start():
    """Start daemon"""
    data_dir = get_data_dir()
    pid_file = data_dir / "daemon.pid"

    if pid_file.exists():
        print("Daemon appears to be running (PID file exists)")
        print("Use 'dere daemon status' to verify")
        sys.exit(1)

    try:
        subprocess.Popen(
            ["dere-daemon"],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            start_new_session=True,
        )
        print("Daemon started")
    except Exception as e:
        print(f"Failed to start daemon: {e}", file=sys.stderr)
        sys.exit(1)


@daemon_app.command()
def stop():
    """Stop daemon"""
    data_dir = get_data_dir()
    pid_file = data_dir / "daemon.pid"

    if not pid_file.exists():
        print("Daemon is not running (no PID file)")
        sys.exit(1)

    try:
        pid = int(pid_file.read_text().strip())
        os.kill(pid, signal.SIGTERM)
        print(f"Sent stop signal to daemon (PID {pid})")
    except ProcessLookupError:
        print("Daemon PID file exists but process not found")
        pid_file.unlink()
    except Exception as e:
        print(f"Failed to stop daemon: {e}", file=sys.stderr)
        sys.exit(1)


@daemon_app.command()
def restart():
    """Restart daemon"""
    data_dir = get_data_dir()
    pid_file = data_dir / "daemon.pid"

    if pid_file.exists():
        try:
            pid = int(pid_file.read_text().strip())
            os.kill(pid, signal.SIGTERM)
            print(f"Stopping daemon (PID {pid})...")
            time.sleep(2)
        except Exception:
            pass

    try:
        subprocess.Popen(
            ["dere-daemon"],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            start_new_session=True,
        )
        print("Daemon restarted")
    except Exception as e:
        print(f"Failed to restart daemon: {e}", file=sys.stderr)
        sys.exit(1)


# Config commands
@config_app.callback(invoke_without_command=True)
def config_callback(ctx: typer.Context):
    """Configuration management"""
    if ctx.invoked_subcommand is None:
        print(ctx.get_help())


@config_app.command("show")
def config_show():
    """Show current configuration"""
    config_path = get_config_dir() / "config.toml"
    if not config_path.exists():
        print(f"Config file not found: {config_path}")
        sys.exit(1)
    print(config_path.read_text())


@config_app.command("path")
def config_path_cmd():
    """Show config file path"""
    print(get_config_dir() / "config.toml")


@config_app.command("edit")
def config_edit():
    """Edit configuration"""
    config_file = get_config_dir() / "config.toml"
    editor = os.environ.get("EDITOR", "nano")
    try:
        subprocess.run([editor, str(config_file)], check=True)
    except Exception as e:
        print(f"Failed to open editor: {e}", file=sys.stderr)
        sys.exit(1)


@typer_app.command()
def version():
    """Show version"""
    print("dere 0.1.0")

File: packages/cli/src/subcommands.ts
import { existsSync } from "node:fs";
import { readFile } from "node:fs/promises";
import { spawn, spawnSync } from "node:child_process";
import { homedir } from "node:os";
import { join } from "node:path";

import { getConfigPath, loadConfig, getDaemonUrlFromConfig } from "@dere/shared-config";

async function resolveDaemonUrl(): Promise<string> {
  const config = await loadConfig();
  return getDaemonUrlFromConfig(config);
}

const MAIN_HELP = `Dere - Personality-layered wrapper for Claude Code

Usage:
  dere [subcommand] [options] [--] [claude args...]

Subcommands:
  daemon    Daemon management
  config    Configuration management
  version   Show version
  -h, --help  Show help
`;

const DAEMON_HELP = `Daemon management

Usage:
  dere daemon status
  dere daemon start
  dere daemon stop
  dere daemon restart
`;

const CONFIG_HELP = `Configuration management

Usage:
  dere config show
  dere config path
  dere config edit
`;

function getDataDir(): string {
  if (process.platform === "darwin") {
    return join(homedir(), "Library", "Application Support", "dere");
  }
  if (process.platform === "win32") {
    const local = process.env.LOCALAPPDATA ?? homedir();
    return join(local, "dere");
  }
  return join(homedir(), ".local", "share", "dere");
}

function pidPath(): string {
  return join(getDataDir(), "daemon.pid");
}

async function daemonStatus(): Promise<void> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 2000);
  try {
    const daemonUrl = await resolveDaemonUrl();
    const response = await fetch(`${daemonUrl}/health`, {
      signal: controller.signal,
    });
    if (!response.ok) {
      console.error("Daemon is not responding correctly");
      process.exit(1);
    }
    const data = (await response.json()) as Record<string, unknown>;
    console.log("Daemon is running");
    console.log(`  DereGraph: ${String(data.dere_graph ?? "unknown")}`);
    console.log(`  Claude auth: ${String(data.claude_auth ?? "unknown")}`);
  } catch {
    console.error("Daemon is not running");
    process.exit(1);
  } finally {
    clearTimeout(timeout);
  }
}

async function daemonStart(): Promise<void> {
  const path = pidPath();
  if (existsSync(path)) {
    console.error("Daemon appears to be running (PID file exists)");
    console.error("Use 'dere daemon status' to verify");
    process.exit(1);
  }

  const child = spawn("bun", ["packages/daemon/src/index.ts"], {
    stdio: "ignore",
    detached: true,
  });
  child.unref();
  console.log("Daemon started");
}

async function daemonStop(): Promise<void> {
  const path = pidPath();
  if (!existsSync(path)) {
    console.error("Daemon is not running (no PID file)");
    process.exit(1);
  }
  try {
    const pid = Number((await readFile(path, "utf-8")).trim());
    if (!Number.isFinite(pid)) {
      throw new Error("Invalid PID");
    }
    process.kill(pid, "SIGTERM");
    console.log(`Sent stop signal to daemon (PID ${pid})`);
  } catch (error) {
    console.error(`Failed to stop daemon: ${String(error)}`);
    process.exit(1);
  }
}

async function daemonRestart(): Promise<void> {
  const path = pidPath();
  if (existsSync(path)) {
    try {
      const pid = Number((await readFile(path, "utf-8")).trim());
      if (Number.isFinite(pid)) {
        process.kill(pid, "SIGTERM");
        console.log(`Stopping daemon (PID ${pid})...`);
        await new Promise((resolve) => setTimeout(resolve, 2000));
      }
    } catch {
      // ignore
    }
  }
  await daemonStart();
  console.log("Daemon restarted");
}

async function configShow(): Promise<void> {
  const configPath = getConfigPath();
  if (!existsSync(configPath)) {
    console.error(`Config file not found: ${configPath}`);
    process.exit(1);
  }
  const text = await readFile(configPath, "utf-8");
  console.log(text);
}

function configPath(): void {
  console.log(getConfigPath());
}

function configEdit(): void {
  const configFile = getConfigPath();
  const editor = process.env.EDITOR ?? "nano";
  const result = spawnSync(editor, [configFile], { stdio: "inherit" });
  if (result.status && result.status !== 0) {
    console.error(`Failed to open editor: ${result.stderr?.toString() ?? ""}`);
    process.exit(result.status);
  }
}

export async function runSubcommand(args: string[]): Promise<void> {
  if (args.length === 0 || args[0] === "--help" || args[0] === "-h") {
    console.log(MAIN_HELP.trim());
    return;
  }

  const [command, ...rest] = args;
  if (command === "version") {
    console.log("dere 0.1.0");
    return;
  }
  if (command === "daemon") {
    const sub = rest[0];
    if (!sub || sub === "--help" || sub === "-h") {
      console.log(DAEMON_HELP.trim());
      return;
    }
    if (sub === "status") {
      await daemonStatus();
      return;
    }
    if (sub === "start") {
      await daemonStart();
      return;
    }
    if (sub === "stop") {
      await daemonStop();
      return;
    }
    if (sub === "restart") {
      await daemonRestart();
      return;
    }
    console.log(DAEMON_HELP.trim());
    process.exit(1);
  }
  if (command === "config") {
    const sub = rest[0];
    if (!sub || sub === "--help" || sub === "-h") {
      console.log(CONFIG_HELP.trim());
      return;
    }
    if (sub === "show") {
      await configShow();
      return;
    }
    if (sub === "path") {
      configPath();
      return;
    }
    if (sub === "edit") {
      configEdit();
      return;
    }
    console.log(CONFIG_HELP.trim());
    process.exit(1);
  }

  console.log(MAIN_HELP.trim());
  process.exit(1);
}
